dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,207 @@
+        ):
+            cached_dirs[dir1["title"]] = dir1["id"]
+        return cached_dirs
+
+    @cached_property
+    def raw_drive(self):
+        from pydrive.auth import GoogleAuth
+        from pydrive.drive import GoogleDrive
+        import logging
+
+        if os.getenv("PYDRIVE_USER_CREDENTIALS_DATA"):
+            with open("credentials.json", "w") as credentials_file:
+                credentials_file.write(
+                    os.getenv("PYDRIVE_USER_CREDENTIALS_DATA")
+                )
+
+        # Supress import error on GoogleAuth warning
+        logging.getLogger("googleapiclient.discovery_cache").setLevel(
+            logging.ERROR
+        )
+
+        GoogleAuth.DEFAULT_SETTINGS["client_config_backend"] = "settings"
+        gauth = GoogleAuth(settings_file=self.gdrive_credentials_path)
+        gauth.CommandLineAuth()
+        return GoogleDrive(gauth)
+
+    @property
+    @ratelimit.sleep_and_retry
+    @ratelimit.limits(calls=10, period=1)
+    def drive(self):
+        return self.raw_drive
+
+    def create_drive_item(self, parent_id, title):
+        item = self.drive.CreateFile(
+            {
+                "title": title,
+                "parents": [{"id": parent_id}],
+                "mimeType": self.FOLDER_MIME_TYPE,
+            }
+        )
+        item.Upload()
+        return item
+
+    def get_drive_item(self, name, parent_id):
+        return next(
+            iter(
+                self.drive.ListFile(
+                    {
+                        "q": "'%s' in parents and trashed=false and title='%s'"
+                        % (parent_id, name)
+                    }
+                ).GetList()
+            ),
+            None,
+        )
+
+    def resolve_remote_file(self, parent_id, path_parts, create):
+        for path_part in path_parts:
+            item = self.get_drive_item(path_part, parent_id)
+            if not item and create:
+                item = self.create_drive_item(parent_id, path_part)
+            elif not item:
+                return None
+            parent_id = item["id"]
+        return item
+
+    def get_path_id_from_cache(self, path_info):
+        file_id = ""
+        parts = path_info.path.split("/")
+        if parts and (parts[0] in self.cached_root_dirs):
+            parent_id = self.cached_root_dirs[parts[0]]
+            file_id = self.cached_root_dirs[parts[0]]
+            parts.pop(0)
+        else:
+            parent_id = path_info.netloc
+        return file_id, parent_id, parts
+
+    def get_path_id(self, path_info, create=False):
+        file_id, parent_id, parts = self.get_path_id_from_cache(path_info)
+
+        if not parts and file_id:
+            return file_id
+
+        file1 = self.resolve_remote_file(parent_id, parts, create)
+        return file1["id"] if file1 else ""
+
+    def exists(self, path_info):
+        return self.get_path_id(path_info) != ""
+
+    def batch_exists(self, path_infos, callback):
```

Comment:
This method is not called by anyone. We used to have it a few months ago, but right now the only remote that uses it is RemoteSSH, and it is only doing that because of more complex SFTP multiplexing. In your current patch, `RemoteBASE.cache_exists` is simply calling `RemoteGDrive.exists()` by default.